internal/notifier/bitbucketserver.go

```diff
@@ -0,0 +1,171 @@
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	"github.com/gdasson/bitbucketv1go"
+)
+
+// Bitbucket is a Bitbucket Server notifier.
+type BitbucketServer struct {
+	ProjectKey      string
+	RepositorySlug  string
+	ProviderUID     string
+	ProviderAddress string
+	Credentials     *bitbucketv1go.Credentials
+	Client          *bitbucketv1go.APIClient
+}
+
+// NewBitbucket creates and returns a new Bitbucket notifier.
+func NewBitbucketServer(providerUID string, addr string, token string, certPool *x509.CertPool, username string, password string) (*BitbucketServer, error) {
+	hst, id, err := parseBitbucketServerGitAddress(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &bitbucketv1go.Credentials{}
+	if len(token) > 0 {
+		c.RestBearerTokenCredentials.Token = token
+	}
+	if len(username) > 0 && len(password) > 0 {
+		c.RestUsernamePasswordCredentials.Username = username
+		c.RestUsernamePasswordCredentials.Password = password
+	}
+	if len(token) == 0 && (len(username) == 0 || len(password) == 0) {
+		return nil, errors.New("invalid credentials, expected to be one of username,password or APIToken")
+	}
+
+	comp := strings.Split(id, "/")
+	if len(comp) != 2 {
+		return nil, fmt.Errorf("invalid repository id %q", id)
+	}
+	projectkey := comp[0]
+	reposlug := comp[1]
+
+	bitbucketConfig := bitbucketv1go.NewConfiguration()
+	bitbucketConfig.BasePath = hst + "/rest"
+	bitbucketConfig.AddDefaultHeader("x-atlassian-token", "no-check")
+	bitbucketConfig.AddDefaultHeader("x-requested-with", "XMLHttpRequest")
+	if certPool != nil {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: certPool,
+			},
+		}
+		hc := &http.Client{Transport: tr}
+		bitbucketConfig.HTTPClient = hc
+	}
+
+	bitbucketClient := bitbucketv1go.NewAPIClient(bitbucketConfig)
+
+	return &BitbucketServer{
+		ProjectKey:      projectkey,
+		RepositorySlug:  reposlug,
+		ProviderUID:     providerUID,
+		Credentials:     c,
+		ProviderAddress: addr,
+		Client:          bitbucketClient,
+	}, nil
+}
+
+// Post Bitbucket build status
+func (b BitbucketServer) Post(ctx context.Context, event eventv1.Event) error {
+	// Skip progressing events
+	if event.HasReason(meta.ProgressingReason) {
+		return nil
+	}
+
+	revString, ok := event.Metadata[eventv1.MetaRevisionKey]
+	if !ok {
+		return errors.New("missing revision metadata")
+	}
+	rev, err := parseRevision(revString)
+	if err != nil {
+		return err
+	}
+	state, err := toBitbucketServerState(event.Severity)
+	if err != nil {
+		return err
+	}
+
+	name, desc := formatNameAndDescription(event)
+	name = name + " [" + desc + "]" //Bitbucket server displays this data on browser. Thus adding description here.
+	id := generateCommitStatusID(b.ProviderUID, event)
+	// key has a limitation of 40 characters in bitbucket api
+	key := sha1String(id)
+
+	if len(b.Credentials.RestUsernamePasswordCredentials.Username) > 0 && len(b.Credentials.RestUsernamePasswordCredentials.Password) > 0 {
```

Comment:
This check is already performed when the struct is created so it's easier to just use
```suggestion
	if b.Credentials.RestUsernamePasswordCredentials != bitbucketv1go.RestUsernamePasswordCredentials{} {
```